src/main/kotlin/calculator/Calculator.kt

```diff
@@ -1,58 +1,26 @@
 package calculator
 
-import calculator.exceptions.NumberSeparationException
 import calculator.exceptions.OperationBetweenNumbersException
 
 object Calculator {
     fun calculate(input: String?) = calculate(InputParser.validateInput(input))
 
     private fun calculate(inputList: List<String>): Double {
-        var currentCalculation = processNumber(inputList.first())
-        var pendingOperation: Operation? = null
+        val numbersList = processNumbers(inputList)
+        val operations: ArrayDeque<Operation> = ArrayDeque(processOperations(inputList))
 
-        for (index in 1..inputList.lastIndex) {
-            if (index % 2 == 0) {
-                // even element - number
-                val number = processNumber(inputList[index])
-                currentCalculation =
-                    performOperation(
-                        operation = requireNotNull(pendingOperation) { "Operation is required on this step" },
-                        a = currentCalculation,
-                        b = number,
-                    )
-            } else {
-                // odd - operation
-                pendingOperation = processOperation(inputList[index])
-            }
+        return numbersList.drop(1).fold(numbersList.first()) { result, number ->
+            operations.removeFirst().apply(a = result, b = number)
         }
```

Comment:
Wow! I didn't mean to turn it into declarative programming, but the syntax looks really cool. 👍 How about making it a bit more readable by adding a line break at each dot?